rig_c_sa/sa.c

```diff
@@ -549,9 +567,70 @@ double sa_get_net_cost(sa_state_t *state, sa_net_t *net) {
 				max_y = net->vertices[i]->y;
 		}
 		
-		// Compute weighted HPWL
-		return sqrt(net->num_vertices) * ((max_x - min_x) + (max_y - min_y)) * net->weight;
+		net->bbox.x1 = min_x;
+		net->bbox.x2 = max_x;
+		net->bbox.y1 = min_y;
+		net->bbox.y2 = max_y;
+	}
+}
+
+double sa_get_net_cost(sa_state_t *state, sa_net_t *net,
+                       sa_bool_t vertices_moved,
+                       int ax, int ay, int bx, int by) {
+	int bbox_width, bbox_height;
+	sa_bool_t recompute_bbox;
+	sa_bool_t bbox_wraps_x;
+	sa_bool_t bbox_wraps_y;
+	
+	// Determine if we can re-use the cached bounding-box information
+	recompute_bbox = sa_true; // Should be overwritten below. Set here defensively.
+	if (net->bbox.x1 < 0) {
+		// Must always recompute if the bbox is not valid.
+		recompute_bbox = sa_true;
+	} else if (!vertices_moved) {
+		// If nothing has moved (and the bbox is valid) we don't need to recompute.
+		recompute_bbox = sa_false;
+	} else {
+		// Things have moved, work out whether any of these changes might change
+		// the bounding box.
+		//
+		// The bounding box may have changed if:
+		//
+		// * A vertex has moved outside the old bounding box.
+		// * A vertex previously 'supporting' one wall of the bounding box is
+		//   moved (possibly making the bounding box smaller).
+		//
+		// Since we don't know which vertices have moved to/from each chip we must
+		// assume the worst case that if either chip coordinate lies on or outside
+		// the net bounding box, the whole bounding box must be recomputed.
+		bbox_wraps_x = net->bbox.x2 < net->bbox.x1;
+		bbox_wraps_y = net->bbox.y2 < net->bbox.y1;
+		recompute_bbox = (
+			(bbox_wraps_x ? (!(ax > net->bbox.x1 || ax < net->bbox.x2) ||
+			                 !(bx > net->bbox.x1 || bx < net->bbox.x2))
+			              : (ax >= net->bbox.x2 || ax <= net->bbox.x1 ||
+			                 bx >= net->bbox.x2 || bx <= net->bbox.x1)) ||
+			(bbox_wraps_y ? (!(ay > net->bbox.y1 || ay < net->bbox.y2) ||
+			                 !(by > net->bbox.y1 || by < net->bbox.y2))
+			              : (ay >= net->bbox.y2 || ay <= net->bbox.y1 ||
+			                 by >= net->bbox.y2 || by <= net->bbox.y1))
+		);
```

Comment:
How easy/beneficial would it be to temporarily translate your co-ordinate system so that (0, 0) matches the bottom left hand corner of the bounding box? That way wrapping disappears as a problem.